### ISSUE TITLE
updating parameters in a batch

### ISSUE THREAD

**Issue by copakahuna** [mitsuba-renderer/mitsuba3#1877]
So batch differentiatoin as in the example
https://mitsuba.readthedocs.io/en/stable/src/inverse_rendering/radiance_field_reconstruction.html#Optimization
should actually work?

eg:
for iteration
  for views
    compute loss
    loss backwards
   opt step

in my little Scene, it only optimizes for the first image :( 
I have 8 reference images, with 8 diferent light positions (ring array) and would like to optimize equally to the lights.

```py
import mitsuba as mi
import drjit as dr
import math
import matplotlib.pyplot as plt
import numpy as np

def initPlot():
    fig, axes = plt.subplots(2, 3, figsize=(12, 8))
    axes = axes.flatten()
    images = []
    for ax in axes:
        img = ax.imshow(np.zeros((10, 10)), cmap='gray', vmin=0, vmax=1)
        ax.axis('off')
        images.append(img)
    plt.tight_layout()
    plt.show(block=False)
    return fig, images


def updatePlot(images, render, ref, extra_rgb):
    diff = dr.abs(render - ref)
    data_list = [render, ref, diff, extra_rgb ]
    clip_range = (0.0, 1.0)
    for im_plot, data in zip(images, data_list):
        img_np = np.array(data)
        img_np = np.clip(img_np, clip_range[0], clip_range[1])
        im_plot.set_data(img_np)
    images[0].figure.canvas.draw()
    images[0].figure.canvas.flush_events()



samples=8
lightIntensityFactor=200
lightIntensity=[1.93252, 1.77156, 1.9246]
lightHeightOffset=15
lightRingRadius=37
diskRadius=2.3
diskAngle=135
LightIntensity=  [3.78724, 3.56801, 3.75235]
lightHeightOffset= 23.8258

mi.set_variant('llvm_ad_rgb')

denominator=8
width=round(5496/denominator);
height=round(3672/denominator);


focalLen=12.0
sensorWidth=13.13
sensorHeight=8.76
distance=31.4


def getPlattenScale(distance):
    w=distance*sensorWidth/focalLen
    h=distance*sensorHeight/focalLen
    initSize=2
    return[w/initSize,h/initSize]
plattenScale=getPlattenScale(distance)


def loadImgIdx(idx, width, height, path, to_gamma=True):
    image_ref = mi.Bitmap(path+'/image'+str(idx)+'.png')
    image_ref = image_ref.convert(mi.Bitmap.PixelFormat.RGB, mi.Struct.Type.Float32, srgb_gamma=True)
    image_ref = image_ref.resample([width, height])
    image_ref = mi.TensorXf(image_ref)
    if not to_gamma:
        return image_ref
    image_ref=dr.srgb_to_linear(image_ref) 
    return image_ref;



def lightDiskTrans(height, radius, idx, discRadius, diskAngle=135):
    angleFrac = 360 / 8
    angle = -idx * angleFrac
    S = mi.ScalarTransform4f.scale(discRadius)
    RToPlane = mi.ScalarTransform4f.rotate([0, 1, 0], 90)
    RAngled = mi.ScalarTransform4f.rotate([0, 0, 1], diskAngle)
    T = mi.ScalarTransform4f.translate([-height, -radius, 0])
    RAroundCenter = mi.ScalarTransform4f.rotate([1, 0, 0], angle)
    return RAroundCenter @ T @ RAngled @ RToPlane @ S

def genScene(idx):
    return {
        "type": "scene",
        "myintegrator": {
            "type": "path",
        },
        "mysensor": {
            "type": "perspective",
            "near_clip": 1.0,
            "far_clip": 1000.0,
            "fov": 2 * math.atan(sensorWidth / (2 * focalLen))*180/math.pi,
            "to_world": mi.ScalarTransform4f.look_at(origin=[0., 0, 0],
                                                    target=[-1, 0, 0], 
                                                    up=[0, 0, 1]),
            "myfilm": {
                "type": "hdrfilm",
                "component_format":"float32",
                "rfilter": {
                    "type": "box"
                },
                "width": width,
                "height": height,
            }, "mysampler": {
                'type': 'multijitter',
            },
        },
        'diskLight': {
            'type': 'disk',
            'to_world': lightDiskTrans(height=lightHeightOffset,radius=lightRingRadius,idx=idx, discRadius=diskRadius ),
            'emitter': {
                'type': 'area',
                'radiance': {
                    'type': 'rgb',
                    'value': [float(x) * lightIntensityFactor for x in lightIntensity],
                }
            }
        },

        "platte": {
            'type': 'rectangle',
            'bsdf': None,
            'to_world': mi.ScalarAffineTransform4f().translate([-distance,0,0]).rotate([0,1,0],90).scale([plattenScale[1], plattenScale[0], 1]).rotate([0.,0.,1.],90),
        },
    }



def getTextureFromVal(val:any, w:int, h:int):
    if isinstance(val, float):
        r = dr.full(mi.TensorXf, val, shape=[h, w, 1])
        return mi.Bitmap(r)

    r = dr.full(mi.TensorXf, val[0], shape=[h, w, 1])
    g = dr.full(mi.TensorXf, val[1], shape=[h, w, 1])
    b = dr.full(mi.TensorXf, val[2], shape=[h, w, 1])
    rgb=dr.concat([r, g, b], axis=-1)
    return mi.Bitmap(rgb)

shared_bsdf = mi.load_dict({
    'type': 'principled',
    'base_color': {
            'type': 'bitmap',
            'raw': True,
            'bitmap': getTextureFromVal([0.5,0.5,0.5], width, height)
    },
    'roughness': {
            'type': 'rgb',
            'value': 1
        },
    'metallic': {
            'type': 'rgb',
            'value': 0
        }
})



scenes = []
for i in range(0, 8):
    scene_dict = genScene(i)
    scene_dict['platte']['bsdf'] = shared_bsdf
    loaded_scene = mi.load_dict(scene_dict)
    scenes.append(loaded_scene)

params = mi.traverse(shared_bsdf)


image_ref=[];
img_folder="/global/path/folder"
for i in range(8):
    image_ref.append(loadImgIdx(i, width, height, img_folder, True))


opt = mi.ad.Adam(lr=0.502)

opt["base_color.data"] = params["base_color.data"]
params.update(opt);


fig, images=initPlot()

for it in range(100):
    ##ONLY optimizes for First image (idx:0)!
    for idx in range(8):
        image = mi.render(scenes[idx], params, spp=samples)
        loss = dr.mean(dr.abs(image - image_ref[idx]))
        dr.backward(loss)
        updatePlot(images, image, image_ref[idx], params['base_color.data'])
    opt.step()
    params.update(opt)
```

**Comment by ziyi-zhang** [comment]
Hi @copakahuna,

The loop should accumulate gradients from all renders before `opt.step()`. But in the code, the first `dr.backward` is likely to have traversed and cleared the graph connecting optimizer parameters and scene parameters. So the remaining iterations cannot propagate derivatives all the way back to optimizer params.  

In general for batching: the [batch sensor](https://mitsuba.readthedocs.io/en/stable/src/generated/plugins_sensors.html#batch-sensor-batch) helps when batching multiple sensors/views. A [RayDataLoader-style setup](https://mitsuba.readthedocs.io/en/latest/src/inverse_rendering/radiance_field_reconstruction.html) (very recently merged) could make sense for ray/pixel minibatching, but the different lighting setups still need to be handled explicitly.

**Comment by copakahuna** [comment]
Thank you so much for your answer @ziyi-zhang  :)
After hours of frustration i think i get the results i would expect, by moving the '''params.update(opt)" call into the inner loop. (for whatever reason, it seems to work)
so maybe this is helpful to anybody :)
```python
for it in range(100):
    for idx in range(8):
        params.update(opt) #<--
        image = mi.render(scenes[idx], params, spp=samples)
        loss = dr.mean(dr.abs(image - image_ref[idx]))
        dr.backward(loss)
        updatePlot(images, image, image_ref[idx], params['base_color.data'])
    opt.step()
```
